runtime/include/QuantumDevice.hpp

```diff
@@ -195,71 +196,70 @@ struct QuantumDevice {
     virtual auto Var(ObsIdType obsKey) -> double = 0;
 
     /**
-     * @brief Compute the probabilities of each computational basis state.
+     * @brief Get the state-vector of a device.
      *
-     * @return `std::vector<double>`
+     * @param state The pre-allocated `MemRefView<complex<double>, 1>`
      */
-    virtual auto Probs() -> std::vector<double> = 0;
+    virtual void State(MemRefView<std::complex<double>, 1> &state) = 0;
 
     /**
-     * @brief Compute the probabilities for a subset of the full system.
-     *
-     * @param wires Wires will restrict probabilities to a subset of the full system
-     *
-     * @return `std::vector<double>`
+     * @brief Compute the probabilities of each computational basis state.
+
+     * @param probs The pre-allocated `MemRefView<double, 1>`
      */
-    virtual auto PartialProbs(const std::vector<QubitIdType> &wires) -> std::vector<double> = 0;
+    virtual void Probs(MemRefView<double, 1> &probs) = 0;
 
     /**
-     * @brief Get the state-vector of a device.
+     * @brief Compute the probabilities for a subset of the full system.
      *
-     * @return `std::vector<std::complex<double>>`
+     * @param probs The pre-allocated `MemRefView<double, 1>`
+     * @param wires Wires will restrict probabilities to a subset of the full system
      */
-    virtual auto State() -> std::vector<std::complex<double>> = 0;
+    virtual void PartialProbs(MemRefView<double, 1> &probs,
+                              const std::vector<QubitIdType> &wires) = 0;
 
     /**
      * @brief Compute samples with the number of shots on the entire wires,
      * returing raw samples.
      *
+     * @param samples The pre-allocated `MemRefView<double, 2>`
      * @param shots The number of shots
-     *
-     * @return `std::vector<double>`
      */
-    virtual auto Sample(size_t shots) -> std::vector<double> = 0;
+    virtual void Sample(MemRefView<double, 2> &samples, size_t shots) = 0;
```

Comment:
This is where specifying what the two different axes mean is important.

Reply:
I've added this to the description of MemRefVew. I'll add it here too. :+1: